### Designer: Maria Gomez

### Description

Two scripts (called by v0.3 of pipeline script countingReads.py) which generate a series of plots describing information generated by the htseq tool. The two scripts create different plots. 

### Code

#### Part One: Pipeline script countsLog_rnaseq.R

The sections of code for the following plots have been modified to remove a suffix from sample names
1. Counts in features
2. Count distribution histogram
3. Count distribution histogram per sample
4. Count distribution lines
5. Count distribution lines nozeros
6. Sample counts correlation heatmap

```R
#!/usr/local/bin/Rscript

suppressMessages(library(RColorBrewer))
suppressMessages(library(matrixStats))
suppressMessages(library(gplots))
suppressMessages(library(ggplot2))
suppressMessages(library(reshape))
suppressMessages(library(GenomicRanges))
suppressMessages(library(rtracklayer))
suppressMessages(library(Rsamtools))
suppressMessages(require(grid))

indir = commandArgs(TRUE)[1]
outdir = commandArgs(TRUE)[2]
#mapping_sum = commandArgs(TRUE)[3]

###############################
# Reading counted reads files #
###############################

#setwd(outdir)
files=list.files(indir, pattern = "_count.txt", full.names=TRUE)

data=read.table(files[1], row.names=1)
colnames(data)=files[1]

for (i in 2:length(files)){
  name=files[i]
  assign(name, read.table(name))
  sample=eval(as.name(name))
  index=match(rownames(data), sample[,1])
  data=cbind(data, sample[index,2])
  colnames(data)[i]=name
}
colnames(data)=gsub('_count.txt','',colnames(data))
colnames(data)=gsub('.*/','',colnames(data))
total_counts=colSums(data)

# Excluding reads not falling in features, ambiguous, too low qual, multiple locations
data.notcounted=data[which(rownames(data)=='__no_feature'):dim(data)[1],]

# Subsetting including reads
data=data[-c(which(rownames(data)=='__no_feature'):dim(data)[1]),]
total.data.counted=colSums(data)

head(data)
# ***************************************************************************************#
# Distribution of counts: From the uniquely mapped reads, we get how many of these were in features, not in feature, ambiguous, and too low quality
n_reads_in_genes=colSums(data)
n_reads_no_feature=t(data.notcounted["__no_feature",])[,1]
n_reads_ambiguous=t(data.notcounted["__ambiguous",])[,1]
n_reads_lowqual=t(data.notcounted["__too_low_aQual",])[,1]

# Using the mapping summary, we get the uniquely mapped reads
#n_reads_total=read.csv(mapping_sum,row.names=1)
#sample_names=rownames(n_reads_total)
#n_reads_total=n_reads_total$Mapped_num
#names(n_reads_total)=sample_names

# Bind matriz
#counting_summ=cbind(n_reads_total, n_reads_in_genes[sample_names], n_reads_no_feature[sample_names], n_reads_ambiguous[sample_names], n_reads_lowqual[sample_names])
#colnames(counting_summ)=c('Uniquely_mapped','In_features', 'Not_in_features', 'Ambiguous', 'Low_aQual')
#write.csv(counting_summ, paste0(outdir,'/counts_in_features.csv'))

# Plot
#table=t(t(counting_summ[,-1]/counting_summ[,1]*100))
#data.melt=melt(table)

##remove _SX from sample names###
#data.melt$X1 = gsub("_S.", "", data.melt$X1) #line added to remove suffix from sample names

#pdf(paste0(outdir,"/counts_in_features.pdf"), width=12)
#ggplot(data.melt, aes(x=X1, y=value, fill=X2, group=X2 )) + geom_bar(stat='identity') + 
#       theme(axis.text.x=element_text(angle=90, hjust=1)) + labs(fill="") + xlab('Sample') + ylab('Percentage')
#dev.off()

# Numbers of genes detected
n_genes=sapply(data,function(x){table(x==0)['FALSE']})
names(n_genes)=gsub('.FALSE','',names(n_genes))
counting_summ=cbind(N_features=n_genes)
write.csv(counting_summ, paste0(outdir,'/Number_of_features_detected.csv'))


# ************************************************ #
# Distribution of counts 
data.melt=suppressMessages(melt(data))
pdf(paste0(outdir,"/countsDistributionHist.pdf"), width=12, height=7)
colnames(data.melt)[1]='Sample'

##remove _SX from sample names###
data.melt$Sample = gsub("_S.", "", data.melt$Sample) #line added to remove suffix from sample names

ggplot(data.melt, aes(x=log(value), fill=Sample )) + geom_bar(binwidth=.2) + 
	theme(legend.text=element_text(size=10), legend.key.size = unit(.78, "cm"))
suppressMessages(dev.off())

pdf(paste0(outdir,"/countsDistributionHist_per_sample.pdf"), width=12, height=7)
colnames(data.melt)[1]='Sample'

##remove _SX from sample names###
data.melt$Sample = gsub("_S.", "", data.melt$Sample) #line added to remove suffix from sample names

ggplot(data.melt, aes(x=log(value), fill=Sample )) + geom_bar(binwidth=.2) + facet_wrap(~Sample) + theme(legend.position='none')
suppressMessages(dev.off())

# Distribution lines
distribution=as.data.frame(matrix(0,ncol=length(files),nrow=14))
colnames(distribution)=colnames(data)
rownames(distribution)=c('0', '1-100', '101-200','200-500','501-1,000', '1,001-5,000', '5,001-10,000', '10,001-20,000', '20,001-30,000', '30,000-40,000', '40,001-50,000', '50,001-100,000', '100,001-50,000', '>500,000')
for (i in 1:length(colnames(data))) {
  sample_name=colnames(data)[i]
  x=data[,sample_name]
  zero=table(x==0)['TRUE']
  bin1=table(x>0 & x<=100)['TRUE']
  bin2=table(x>100 & x<=200)['TRUE']
  bin3=table(x>200 & x<=500)['TRUE']
  bin4=table(x>500 & x<=1000)['TRUE']
  bin5=table(x>1000 & x<=5000)['TRUE']
  bin6=table(x>5000 & x<=10000)['TRUE']
  bin7=table(x>10000 & x<=20000 )['TRUE']
  bin8=table(x>20000 & x<=30000 )['TRUE']
  bin9=table(x>30000 & x<=40000 )['TRUE']
  bin10=table(x>40000 & x<=50000 )['TRUE']
  bin11=table(x>50000 & x<=100000 )['TRUE']
  bin12=table(x>100000 & x<=500000 )['TRUE']
  bin13=table(x>500000)['TRUE']
  sample=c(zero,bin1,bin2,bin3,bin4,bin5,bin6,bin7,bin8,bin9,bin10,bin11,bin12,bin13)
  sample[is.na(sample)]=0
  distribution[,sample_name]=sample
}
distribution=distribution/colSums(distribution)*100
distribution.melt=melt(cbind(bins=factor(row.names(distribution), levels=row.names(distribution)), distribution))

##remove _SX from sample names###
distribution.melt$variable = gsub("_S.", "", distribution.melt$variable) #line added to remove suffix from sample names

pdf(paste0(outdir,'/countDistribution_lines.pdf'), width=12)
ggplot(distribution.melt,aes(x=bins,y=value,color=variable,group=variable)) + geom_path() + 
  xlab('Number of counts') + ylab('Percentage of genes') + 
  scale_y_continuous(breaks=seq(0,70,10), labels=c('0%','10%','20%','30%','40%','50%','60%','70%')) + 
  scale_colour_discrete(name='Sample') + 
  theme(axis.text.x=element_text(angle=90), axis.text.y=element_text(),legend.text=element_text(size=10), legend.key.size = unit(.78, "cm"))
dev.off()

# Distribution lines removing all zeros
data_filtered=data[!rowSums(data)==0,]
distribution=as.data.frame(matrix(0,ncol=length(files),nrow=14))
colnames(distribution)=colnames(data_filtered)
rownames(distribution)=c('0', '1-100', '101-200','200-500','501-1,000', '1,001-5,000', '5,001-10,000', '10,001-20,000', '20,001-30,000', '30,000-40,000', '40,001-50,000', '50,001-100,000', '100,001-50,000', '>500,000')
for (i in 1:length(colnames(data_filtered))) {
  sample_name=colnames(data_filtered)[i]
  x=data_filtered[,sample_name]
  zero=table(x==0)['TRUE']
  bin1=table(x>0 & x<=100)['TRUE']
  bin2=table(x>100 & x<=200)['TRUE']
  bin3=table(x>200 & x<=500)['TRUE']
  bin4=table(x>500 & x<=1000)['TRUE']
  bin5=table(x>1000 & x<=5000)['TRUE']
  bin6=table(x>5000 & x<=10000)['TRUE']
  bin7=table(x>10000 & x<=20000 )['TRUE']
  bin8=table(x>20000 & x<=30000 )['TRUE']
  bin9=table(x>30000 & x<=40000 )['TRUE']
  bin10=table(x>40000 & x<=50000 )['TRUE']
  bin11=table(x>50000 & x<=100000 )['TRUE']
  bin12=table(x>100000 & x<=500000 )['TRUE']
  bin13=table(x>500000)['TRUE']
  sample=c(zero,bin1,bin2,bin3,bin4,bin5,bin6,bin7,bin8,bin9,bin10,bin11,bin12,bin13)
  sample[is.na(sample)]=0
  distribution[,sample_name]=sample
}
distribution=distribution/colSums(distribution)*100
distribution.melt=melt(cbind(bins=factor(row.names(distribution), levels=row.names(distribution)), distribution))

##remove _SX from sample names###
distribution.melt$variable = gsub("_S.", "", distribution.melt$variable) #line added to remove suffix from sample names

pdf(paste0(outdir,'/countDistribution_lines_nozeros.pdf'), width=12)
ggplot(distribution.melt,aes(x=bins,y=value,color=variable,group=variable)) + geom_path() + 
  xlab('Number of counts') + ylab('Percentage of genes') + 
  scale_y_continuous(breaks=seq(0,70,10), labels=c('0%','10%','20%','30%','40%','50%','60%','70%')) + 
  scale_colour_discrete(name='Sample') + 
  theme(axis.text.x=element_text(angle=90), axis.text.y=element_text(),legend.text=element_text(size=10), legend.key.size = unit(.78, "cm"))
dev.off()


### heatmap correlation samples based on count profile
data_norm=t(t(data)/rowSums(t(data)))
colnames(data_norm) #line added to allow confirmation that suffix removal has worked
colnames(data_norm) <- gsub("_S.", "", colnames(data_norm)) #line added to remove suffix from sample names
colnames(data_norm) #line added to allow confirmation that suffix removal has worked
pdf(paste0(outdir,'/sample_counts_correlation_heatmap.pdf'))
heatmap.2(cor(data_norm), scale=c('none'), margins=c(10,10),density.info='density', trace='none')
dev.off()

```

#### Part Two: Pipeline script library_proportion.R

The sections of code for the following plots have been modified to remove a suffix from sample names
1. Biotype proportion
2. Top expressed genes
3. Top expressed genes protein coding

```R
#!/usr/local/bin/Rscript

suppressMessages(library(gplots))
suppressMessages(library(rtracklayer))
suppressMessages(library(reshape))
suppressMessages(library(ggplot2))

indir = commandArgs(TRUE)[1]
outdir = commandArgs(TRUE)[2]
gtf.file= commandArgs(TRUE)[3]

dir.create(outdir, showWarnings = FALSE)

# Reading counted reads files #
#-----------------------------#

files=list.files(indir, pattern = "_count.txt", full.names=TRUE)

data=read.table(files[1], row.names=1)
colnames(data)=files[1]

for (i in 2:length(files)){
  name=files[i]
  assign(name, read.table(name))
  sample=eval(as.name(name))
  index=match(rownames(data), sample[,1])
  data=cbind(data, sample[index,2])
  colnames(data)[i]=name
}
colnames(data)=gsub('_count.txt','',colnames(data))
colnames(data)=gsub('.*/','',colnames(data))

# Exclude reads not mapping to features, ambiguous, too low qual, multiple locations
data.notcounted=data[which(rownames(data)=='__no_feature'):dim(data)[1],]

# Extract reads mapping to features
data=data[-c(which(rownames(data)=='__no_feature'):dim(data)[1]),]
total.data.counted=colSums(data)

# Read GTF File
GTF <- import.gff(gtf.file, format="gtf", feature.type="gene")
df=as.data.frame(GTF)
df=subset(df,select=c(seqnames,start,end,width,strand,gene_id,gene_name,gene_biotype))
rownames(df)=df$gene_id
df=df[rownames(data),]
df$gene_name[is.na(df$gene_name)]=df$gene_id[is.na(df$gene_name)]

# Library proportion of different biotypes #
#------------------------------------------
type=unique(df$gene_biotype)
for (i in 1:length(type)){ 
	name=type[i]
	assign(name,colSums(data[which(df$gene_biotype==type[i]),])/colSums(data[,]))
}

matrix=matrix(ncol=ncol(data),nrow=length(type))
rownames(matrix)=type
colnames(matrix)=colnames(data)
for (i in 1:length(type)){
	matrix[type[i],]=eval(as.name(type[i]))
}
#matrix=matrix[which(rowSums(matrix)==0),]
matrix.melt=melt(matrix)
colnames(matrix.melt)[2]='Sample'

##remove _SX from sample names###
matrix.melt$Sample = gsub("_S.", "", matrix.melt$Sample) #line added to remove suffix from sample names

pdf(paste0(outdir,'/biotype_proportion.pdf'),width=10, height=7)
ggplot(matrix.melt, aes(x=X1, y=value, colour=Sample, shape=Sample)) + geom_point(size=1) + 
theme(axis.text.x=element_text(), axis.text=element_text(size=10, colour='black'), axis.title.x=element_blank(), legend.text=element_text(size=8), 
   panel.grid.major=element_line(colour='bisque3', linetype='dashed', size=.3), panel.background=element_rect(colour='white')) + 
scale_shape_manual(values=0:ncol(data)) + ylab('Library Proportion') + xlab('') +
coord_flip() 
dev.off()

# Library proportion of different genes #
#---------------------------------------#
prop=t(t(data)/total.data.counted)
prop.adjusted=prop/df$width

# Ignore ribosomal protein genes
#rps=df[grep('Rps',df$gene_name),]$gene_id
#rpl=df[grep('Rpl',df$gene_name),]$gene_id
#mt=df[grep('mt-',df$gene_name),]$gene_id
#remove.genes=c(rps,rpl,mt)
#prop.adjusted.noribosomalprotein=prop.adjusted[!rownames(prop.adjusted)%in%remove.genes,]

q=min(apply(prop.adjusted,2,function(x){quantile(x, .999)}))
genes_most_expressed=sapply(1:ncol(prop.adjusted),function(x){which(prop.adjusted[,x]>.0000002)})
genes_most_expressed=sapply(1:ncol(prop.adjusted),function(x){which(prop.adjusted[,x]>q)})
#print(head(genes_most_expressed))
#names(genes_most_expressed)=colnames(prop.adjusted)
#genes_most_expressed=lapply(genes_most_expressed, function(x){df[names(x),]})
#genes_most_expressed=lapply(names(genes_most_expressed),function(x){cbind(genes_most_expressed[[x]],prop=prop[rownames(genes_most_expressed[[x]]),x])})
#names(genes_most_expressed)=colnames(prop.adjusted)

genes=unique(names(unlist(genes_most_expressed)))
prop.top=prop[genes,]
prop.top=cbind(prop.top,subset(df[genes,],select=c(seqnames,strand,gene_name,gene_biotype)))

#prop.top$variable=factor(prop.top$variable)
prop.top=melt(prop.top)
prop.top$variable=factor(prop.top$variable)
colnames(prop.top)[5]='Sample' #line added to set up for suffix removal

##remove _SX from sample names###
prop.top$Sample = gsub("_S.", "", prop.top$Sample) #line added to remove suffix from sample names

pdf(paste0(outdir,'/top_expressed_genes.pdf'),width=12, height=7)
ggplot(prop.top,aes(x=gene_name,y=value,group=Sample,color=Sample, shape=Sample)) +
geom_point(size=1.5) + facet_wrap(~gene_biotype, scales='free_x')+ xlab('') + ylab('Library proportion') + theme(axis.text=element_text(size=6),axis.text.x=element_text(angle=90), axis.title=element_text(size=10), legend.text=element_text(size=8), legend.title=element_text(size=8)) + scale_shape_manual(values=0:ncol(data)) 
#ggplot(group, colour, shape) elements changed to 'Sample' as part of suffix removal 
dev.off()
#legend.key.height=unit(.3, 'cm')

prop.top=prop.top[prop.top$gene_biotype=='protein_coding',]
prop.top$gene_name=factor(prop.top$gene_name,levels=sort(unique(prop.top$gene_name), decreasing=TRUE))
colnames(prop.top)[5]='Sample'

##remove _SX from sample names###
prop.top$Sample = gsub("_S.", "", prop.top$Sample) #line added to remove suffix from sample names

pdf(paste0(outdir,'/top_expressed_genes_protein_coding.pdf'),width=12, height=10)
ggplot(prop.top,aes(x=gene_name,y=value,group=Sample,color=Sample, shape=Sample)) +
geom_point(size=3) + 
facet_wrap(~gene_biotype, scales='free_x')+ xlab('') + ylab('Library proportion') + 
theme(axis.text=element_text(size=8, color='black'),axis.text.x=element_text(), 
      axis.title=element_text(size=10), 
      legend.text=element_text(size=8),  legend.title=element_text(size=8), 
      panel.grid.major=element_line(colour='bisque3', linetype='dashed', size=.3), 
      panel.background=element_rect(colour='white')) + 
scale_shape_manual(values=0:ncol(data)) + 
coord_flip()
dev.off()
#legend.key.height=unit(.3, 'cm')

```
